js_tests/basic.test.ts

```diff
@@ -0,0 +1,174 @@
+import { RpcProvider, Account, Contract, CallData, cairo } from "starknet";
+import { RPC_URL, SIGNER_PRIVATE, SIGNER_PUBLIC } from "./constant";
+import { readContractSierra, readContractCasm } from "./utils";
+import { hash } from "starknet";
+
+/**
+ * Test order configuration
+ * Each test is defined with a name and corresponding function
+ */
+const testOrder = [
+  { name: "Declare Contract", fn: declareContract },
+  { name: "Deploy Contract", fn: deployContract },
+  { name: "Deploy Account", fn: deployAccount },
+  { name: "Fund Transfer", fn: transferFunds },
+];
+
+/**
+ * Test context interface
+ */
+interface TestContext {
+  provider: RpcProvider;
+  account: Account;
+  nonce: number;
+}
+
+describe("Starknet Contract Tests", () => {
+  let provider: RpcProvider;
+  let account: Account;
+  let nonce: number;
+
+  beforeAll(async () => {
+    // Initialize provider and account
+    provider = new RpcProvider({ nodeUrl: RPC_URL });
+    account = new Account(provider, SIGNER_PUBLIC, SIGNER_PRIVATE);
+    nonce = Number(await account.getNonce());
+  });
+
+  // Run tests in specified order
+  testOrder.forEach(({ name, fn }) => {
+    test(name, async () => {
+      const result = await fn({ provider, account, nonce });
+      nonce++; // Increment nonce after each test
+
+      // TODO: this is a workaround to wait for the block to be produced
+      // given we have block time of 3 seconds, we wait for 5 seconds so that
+      // nonce gets updated. Remove this when we have a way to get the updated nonce
+      // when there is a txn already in the pool from same account
+      await new Promise((resolve) => setTimeout(resolve, 5000));
+    });
+  });
+});
+
+/**
+ * Declares a contract on the Starknet network
+ *
+ * @param {TestContext} context - The test context
+ * @returns {Promise<{classHash: string}>} The declared class hash
+ */
+async function declareContract({ provider, account, nonce }: TestContext) {
+  const sierra = readContractSierra("madara_contracts_HelloStarknet");
+  const casm = readContractCasm("madara_contracts_HelloStarknet");
+
+  const declareResponse = await account.declare(
+    {
+      contract: sierra,
+      casm: casm,
+    },
+    {
+      nonce,
+    },
+  );
+
+  await provider.waitForTransaction(declareResponse.transaction_hash);
+
+  // Add assertion to check if the class hash is valid
+  expect(declareResponse.class_hash).toBeTruthy();
+}
+
+/**
+ * Deploys a contract to the Starknet network
+ *
+ * @param {TestContext} context - The test context
+ */
+async function deployContract({ provider, account, nonce }: TestContext) {
+  const deployResult = await account.deploy(
+    {
+      classHash: hash.computeContractClassHash(
+        readContractSierra("madara_contracts_HelloStarknet"),
+      ),
+    },
+    {
+      nonce,
+    },
+  );
+
+  // Add assertion to check if the contract address is valid
+  expect(deployResult.contract_address[0]).toBeTruthy();
```

Comment:
we can maybe call get class hast at and check if it's registered on chain

Reply:
technically we don't need it because getClassAt is the first call that is made when trying to deploy a contract.

here is the calls that are made when trying to deploy a contract:
```
[2024-09-27 11:47:01 HTTP] 🌐 starknet_V0_7_1_getClassAt 200 112069 bytes - 3.492ms
[2024-09-27 11:47:02 HTTP] 🌐 starknet_V0_7_1_estimateFee 200 162 bytes - 8.587ms
[2024-09-27 11:47:02 HTTP] 🌐 starknet_V0_7_1_addInvokeTransaction 200 123 bytes - 6.325ms
```

Reply:
resolved